### PR TITLE
Draft: Adding `resolve` to options passed in from compose

### PR DESCRIPTION
**Pull request by dzearing**
This adds the "mergeProps" logic from my compose work to the react-compose package.

This allows us to standardize how we resolve shorthand, as prop, and distribute to the given slots.

Added lots of tests. Still needs much more!

Note: I am still finalizing the PR, but want to get a build going.

### CODE REVIEW

File: packages/react-compose/src/createClassResolver.test.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  it('can resolve slot classes', () => {
+    expect(classResolver({}, slots)).toEqual({
+      root: 'root',
+      slot1: 'slot1',
+      slot2: 'slot2',
+    });
+  });
+
+  it('can resolve modifiers', () => {
+    expect(classResolver({ primary: true }, slots)).toEqual({
```
Comment (xugao) [inline packages/react-compose/src/createClassResolver.test.ts:23]: with our scss conversions, we are actually adding modifier classes onto every slot. that way the scss can be written in a cleaner way.

File: packages/react-button/src/components/Button/Button.tsx [hunk 2]
```diff
@@ -1,12 +1,10 @@
...
 import { ButtonProps } from './Button.types';
+import { compose, createClassResolver } from '@fluentui/react-compose';
+import * as classes from './Button.scss';
 
-export const Button = compose<'button', {}, {}, ButtonProps, ButtonProps>(ButtonBase, {
-  classes,
-  slots: {
-    icon: 'div',
-  },
+export const Button = compose<'button', ButtonProps, ButtonProps, {}, {}>(ButtonBase, {
```
Comment (xugao) [inline packages/react-button/src/components/Button/Button.tsx:6]: let's add a flamegrill test for this

File: packages/react-compose/src/createClassResolver.test.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    expect(classResolver({ size: 'small' }, slots)).toEqual({
+      root: 'root small',
+      slot1: 'slot1',
+      slot2: 'slot2',
+    });
+  });
+
+  it('can resolve mixed content, including a className in props', () => {
+    // Can resolve
+    expect(classResolver({ className: 'foo', primary: true, size: 'medium' }, slots)).toEqual({
```
Comment (xugao) [inline packages/react-compose/src/createClassResolver.test.ts:41]: lets add test for when `state` has properties that are not mapped to classes:
```
classResolver({ className: 'foo', primary: true, size: 'medium', bar: 'bar' }, slots)
```

File: packages/react-compose/src/mergeComposeOptions.ts [hunk 4]
```diff
@@ -32,74 +15,83 @@ export const defaultComposeOptions: ComposePreparedOptions = {
...
-  return {
+  const preparedOptions: Partial<ComposePreparedOptions<TElementType, TProps, TState>> = {
+    // remove
     className: inputOptions.className || parentOptions.className,
-    classes: mergeClasses(parentOptions.classes, inputOptions.classes),
+
+    classes: [...parentOptions.classes, inputOptions.classes],
 
     displayName: inputOptions.displayName || parentOptions.displayName,
+    // remove
```
Comment (dzearing) [inline packages/react-compose/src/mergeComposeOptions.ts:44]: remove comments or code

File: packages/react-compose/src/createOptionsResolver.ts [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+//   // Mix in classes as needed
+//   mergeClassesIntoSlotProps(options.classes, state, slots, slotProps);
+
+//   return {
+//     state,
+//     slots,
+//     slotProps,
+//   } as ComponentHookResult<TState, TSlots, TSlotProps>;
+// }
```
Comment (khmakoto) [inline packages/react-compose/src/createOptionsResolver.ts:269]: Delete this?

File: packages/react-compose/src/mergeClasses.test.ts [hunk 6]
```diff
@@ -20,4 +20,21 @@ describe('mergeClasses', () => {
...
+  // it('can merge class functions', () => {
+  //   type ExampleState = { foo: string; bar: string };
+  //   const result = mergeClasses(
+  //     (state: ExampleState) => ({ root: 'a', foo: state.foo }),
+  //     (state: ExampleState) => ({ root: 'b', bar: state.bar }),
+  //   );
+
+  //   expect(result).toBeTruthy();
+  //   expect(result({ foo: 'foo', bar: 'bar' })).toEqual({ root: 'a b', foo: 'foo', bar: 'bar' });
+  // });
```
Comment (khmakoto) [inline packages/react-compose/src/mergeClasses.test.ts:33]: Removing?

File: packages/react-compose/src/mergeClasses.test.ts [hunk 7]
```diff
@@ -20,4 +20,21 @@ describe('mergeClasses', () => {
...
+
+  it('can merge an object and a function', () => {});
+
+  it('can merge a function and an object', () => {});
+
+  it('can merge mixed content', () => {});
```
Comment (khmakoto) [inline packages/react-compose/src/mergeClasses.test.ts:39]: We're missing the implementation of this.

File: packages/react-compose/src/compose.ts [hunk 8]
```diff
@@ -1,37 +1,66 @@
...
-      });
+      // tslint:disable-next-line:no-any
+      return composeOptions.render(props, ref as any, (composeOptions as unknown) as ComposePreparedOptions);
+      // return composeOptions.render(props, (ref as unknown) as React.Ref<HTMLDivElement>, {
+      //   ...composeOptions,
+      //   slots: {
+      //     ...composeOptions.slots,
+      //     __self: Component,
+      //   },
+      // });
```
Comment (xugao) [inline packages/react-compose/src/compose.ts:50]: revert? need to keep `__self`: https://github.com/microsoft/fluentui/pull/13115

File: packages/react-compose/src/createOptionsResolver.ts [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    map[key].push(value);
+  }
+}
+
+function assignToMapObject(map: Record<string, {}>, key: string, value: {}) {
+  if (value) {
+    if (!map[key]) {
+      map[key] = {};
+    }
+    Object.assign(map[key], value);
```
Comment (xugao) [inline packages/react-compose/src/createOptionsResolver.ts:67]: do we have polyfill for object.assign?

File: packages/react-compose/src/createOptionsResolver.ts [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    slots.root = state.as || slots.root || 'div';
+
+    // Resolve unrecognized props
+    assignToMapObject(slotProps, 'root', getNativeElementProps(state.as, state));
+
+    // Resolve slotProps/slots from state
+    resolveSlotProps(result);
+
+    // Resolve classes
+    resolveClasses(options.classes, result);
```
Comment (xugao) [inline packages/react-compose/src/createOptionsResolver.ts:48]: move these to separate files and add tests for each individually?

File: packages/react-compose/src/createOptionsResolver.ts [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+      slotProps[key].className = classMap[key].join(' ');
+    }
+  }
+}
+
+/**
+ * NOTE! THIS FILE IS TEMPORARY AND SHOULD BE DELETED ONCE IT HAS MOVED TO `@fluentui/react-compose`.
+ */
+
```
Comment (xugao) [inline packages/react-compose/src/createOptionsResolver.ts:150]: can remove below

File: packages/react-compose/src/mergeComposeOptions.ts [hunk 12]
```diff
@@ -32,74 +15,83 @@ export const defaultComposeOptions: ComposePreparedOptions = {
...
-          };
-        }
-      });
-
-      return mergedSlotProps;
-    }, {});
-
-  return {
+  const preparedOptions: Partial<ComposePreparedOptions<TElementType, TProps, TState>> = {
+    // remove
```
Comment (khmakoto) [inline packages/react-compose/src/mergeComposeOptions.ts:38]: Are these TODOs for later PRs?

File: packages/react-compose/src/types.ts [hunk 13]
```diff
@@ -34,60 +40,85 @@ export interface ComponentWithAs<E extends React.ElementType = 'div', P = {}> ex
...
+  ref: React.Ref<TElementType extends keyof HTMLElementTagNameMap ? HTMLElementTagNameMap[TElementType] : TElementType>,
   composeOptions: ComposePreparedOptions,
 ) => React.ReactElement | null;
 
-export type ComposeOptions<InputProps = {}, InputStylesProps = {}, ParentStylesProps = {}> = {
+export type ComposeOptions<
+  TInputProps = {},
+  TInputStylesProps = {},
+  TParentStylesProps = {},
+  TInputState = TInputProps
```
Comment (khmakoto) [inline packages/react-compose/src/types.ts:70]: What are we using `TInputState` on here?

File: packages/react-compose/tslint.json [hunk 14]
```diff
@@ -1,4 +1,6 @@
 {
   "extends": ["@uifabric/tslint-rules"],
-  "rules": {}
+  "rules": {
+    "interface-name": [true, "never-prefix"]
```
Comment (mnajdova) [inline packages/react-compose/tslint.json:4]: 👍

File: packages/react-compose/src/types.ts [hunk 15]
```diff
@@ -80,21 +95,29 @@ export type GenericDictionary = Record<string, any>;
...
   slots: Record<string, React.ElementType> & { __self: React.ElementType };
-  mapPropsToSlotPropsChain: ((props: Props) => Record<string, object>)[];
-
+  mapPropsToSlotPropsChain: ((props: TProps) => Record<string, object>)[];
   resolveSlotProps: <P>(props: P) => Record<string, object>;
-  shorthandConfig: ShorthandConfig<Props>;
+
+  shorthandConfig: ShorthandConfig<TProps>;
+
+  resolve: (
```
Comment (mnajdova) [inline packages/react-compose/src/types.ts:116]: We should not have both `resolveSlotProps` and `resolveProps`. `compose` is already confusing enough

File: packages/react-compose/etc/react-compose.api.md [hunk 16]
```diff
@@ -10,64 +10,70 @@ import * as React from 'react';
...
 // @public (undocumented)
-export type ComposedComponent<P = {}> = React.FunctionComponent<P> & {
-    fluentComposeConfig: ComposePreparedOptions;
+export type ComposedComponent<TElementType extends React.ElementType<any> = 'div', TProps = {}, TState = TProps> = React.FunctionComponent<TProps> & {
+    fluentComposeConfig: ComposePreparedOptions<TElementType, TProps, TState>;
 };
 
 // @public (undocumented)
-export type ComposeOptions<InputProps = {}, InputStylesProps = {}, ParentProps = {}, ParentStylesProps = {}> = {
+export type ComposeOptions<TInputProps = {}, TInputStylesProps = {}, TParentProps = {}, TParentStylesProps = {}> = {
```
Comment (mnajdova) [inline packages/react-compose/etc/react-compose.api.md:39]: Would be better to stage the generics rename as a separate PR, just for easier reviewing :)

File: packages/react-compose/etc/react-compose.api.md [hunk 17]
```diff
@@ -10,64 +10,70 @@ import * as React from 'react';
...
     className?: string;
-    classes?: ClassDictionary;
+    classes?: ClassDictionary | ((state: GenericDictionary, slots: GenericDictionary) => ClassDictionary);
     displayName?: string;
-    mapPropsToStylesProps?: (props: ParentStylesProps & InputProps) => InputStylesProps;
-    handledProps?: (keyof InputProps | 'as')[];
+    defaultProps?: TInputProps;
```
Comment (mnajdova) [inline packages/react-compose/etc/react-compose.api.md:43]: If this is the new pattern we should update all components that use `compose`

File: packages/react-button/src/components/Button/useButton.ts [hunk 18]
```diff
@@ -1,11 +1,9 @@
...
 import { ComposePreparedOptions } from '@fluentui/react-compose';
 
 /**
  * The useButton hook processes the Button component props and returns
  * state, slots, and slotProps for consumption by the component.
  * @param props
  */
-export const useButton = (props: ButtonProps, options: ComposePreparedOptions) =>
-  mergeProps<ButtonProps, ButtonSlots, ButtonSlotProps>(props, options);
+export const useButton = (props: ButtonProps, options: ComposePreparedOptions) => options.resolve(props);
```
Comment (mnajdova) [inline packages/react-button/src/components/Button/useButton.ts:9]: Nothing buttonish about this hook anymore :D

File: packages/react-compose/src/compose.ts [hunk 19]
```diff
@@ -1,37 +1,66 @@
...
-          __self: Component,
-        },
-      });
+      // tslint:disable-next-line:no-any
+      return composeOptions.render(props, ref as any, (composeOptions as unknown) as ComposePreparedOptions);
+      // return composeOptions.render(props, (ref as unknown) as React.Ref<HTMLDivElement>, {
+      //   ...composeOptions,
+      //   slots: {
+      //     ...composeOptions.slots,
+      //     __self: Component,
```
Comment (mnajdova) [inline packages/react-compose/src/compose.ts:48]: We need to preserve the `__self` it us used at least in the Menu component

File: packages/react-compose/src/compose.ts [hunk 20]
```diff
@@ -1,37 +1,66 @@
...
+ * The `compose` utility allows React components to be created using a set of default config settings. Resulting
+ * component can be "recomposed" by calling `compose` with them as the first argument, attaching config overrides
+ * in the second argument. Settings are merged correctly and provided to the component render function.
+ *
+ * @param input - render function `(props, ref, config) => JSX.Element` or previously composed component.
+ * @param inputOptions - options for rendering the component.
+ */
+function compose<
+  TElementType extends React.ElementType,
+  TProps = {},
```
Comment (mnajdova) [inline packages/react-compose/src/compose.ts:22]: Would be better to stage the generics rename as a separate PR, just for easier reviewing :)

File: packages/react-compose/src/compose.ts (hunk 19 above)
Comment (dzearing) [reply packages/react-compose/src/compose.ts:48]: got it in my reworked pr

File: packages/react-button/src/components/Button/useButton.ts (hunk 18 above)
Comment (dzearing) [reply packages/react-button/src/components/Button/useButton.ts:9]: leaving it there for future updates.

**Comment by dzearing** [comment]
Closing in light of the other split out pr

**Pull request closed without merging**